### PR TITLE
Made the OCI8Statement class final

### PR DESCRIPTION
**Pull request by morozov**
|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | yes
| Fixed issues | #3818
| Depends on | #3809

The `OCI8Statement::convertPositionalToNamedPlaceholders()` method is extracted to an internal utility class in order to enable making the class final. The corresponding test methods are extracted from `UtilTest` and `OCI8StatementTest` to the dedicated test class. 

`OCI8StatementTest` has been removed since it doesn't contain any more tests which is a good sign: we're not testing any driver-specific implementation details on the driver class itself.

### CODE REVIEW

File: UPGRADE.md [hunk 1]
```diff
@@ -1,5 +1,13 @@
 # Upgrade to 3.0
 
+## BC BREAK: `OCI8Statement::convertPositionalToNamedPlaceholders()` is removed.
+
+The `OCI8Statement::convertPositionalToNamedPlaceholders()` method has been extracted to an internal utility class.
+
+## BC BREAK: Dropped handling of one-based numeric arrays of parameters in `Statement::execute()`
+
+The statement implementations no longer detect whether `$params` is a zero- or one-based array. A zero-based numeric array is expected.
```
Comment (Ocramius) [inline UPGRADE.md]: We should probably start adding `@psalm-param list<T>` for these

**Review by Ocramius — APPROVED** [review]
LGTM 👍

File: UPGRADE.md (hunk 1 above)
Comment (morozov) [reply UPGRADE.md]: Should we add Psalm to the build pipeline first?

Comment (Ocramius) [reply UPGRADE.md]: Sure thing - it's a big amount of work to get it to "like" our codebase though 😅

**Pull request merged**